sharding/utils/marshal.go

```diff
@@ -45,149 +48,140 @@ func ConvertFromRawBlob(blob *RawBlob, i interface{}) error {
 
-		// Terminal chunk has its indicator byte added, chunkDataSize*chunksNumber refers to the total size of the blob
-		tempBody = append(tempBody,
-			append(indicatorByte,
-				cb.data[(chunksNumber-1)*chunkDataSize:chunkDataSize*chunksNumber]...)...)
+		for j := 0; j < numChunks; j++ {
+			var terminalLength int
 
-		return tempBody, nil
+			// if non-terminal chunk
+			if j != numChunks-1 {
+				terminalLength = int(chunkDataSize)
 
-	}
+				// append indicating byte with just the length bits
+				returnData = append(returnData, byte(0))
+			} else {
+				terminalLength = getTerminalLength(len(rawBlob.data))
 
-	// This loop loops through all non-terminal chunks and add a indicator byte
-	// of 00000000, each chunk is created by appending the indcator byte
-	// to the data chunks. The data chunks are separated into sets of 31.
-	for i := int64(1); i <= chunksNumber; i++ {
-		tempBody = append(tempBody,
-			append(indicatorByte,
-				cb.data[(i-1)*chunkDataSize:i*chunkDataSize]...)...)
-	}
-	// Appends indicator bytes to terminal-chunks , and if the index of the chunk
-	// delimiter is non-zero adds it to the chunk. Also pads empty bytes to
-	// the terminal chunk.chunkDataSize*chunksNumber refers to the total
-	// size of the blob. finalchunkIndex refers to the index of the last data byte.
-	indicatorByte[0] = byte(terminalLength)
-	if cb.flags.skipEvmExecution {
-		indicatorByte[0] |= (1 << 7)
-	}
-	tempBody = append(tempBody,
-		append(indicatorByte,
-			cb.data[chunkDataSize*chunksNumber:length]...)...)
+				indicatorByte := byte(terminalLength)
+				// include skipEvm flag if true
+				if rawBlob.flags.skipEvmExecution {
+					indicatorByte = indicatorByte | skipEvmBits
+				}
 
-	emptyBytes := make([]byte, (chunkDataSize - terminalLength))
-	tempBody = append(tempBody, emptyBytes...)
+				returnData = append(returnData, indicatorByte)
+			}
 
-	return tempBody, nil
+			// append data bytes
+			chunkStart := j * int(chunkDataSize)
+			chunkEnd := chunkStart + terminalLength
+			blobSlice := rawBlob.data[chunkStart:chunkEnd]
+			returnData = append(returnData, blobSlice...)
+
+			// append filler bytes, if necessary
+			if terminalLength != int(chunkDataSize) {
+				numFillerBytes := numChunks*int(chunkDataSize) - len(rawBlob.data)
+				fillerBytes := make([]byte, numFillerBytes)
+				returnData = append(returnData, fillerBytes...)
+			}
+		}
+	}
 
+	return returnData, nil
 }
 
-// Serialize takes a set of blobs and converts them to a single byte array.
-func Serialize(rawblobs []*RawBlob) ([]byte, error) {
-	length := int64(len(rawblobs))
-
-	serialisedData := []byte{}
-
-	//Loops through all the blobs and serializes them into chunks
-	for i := int64(0); i < length; i++ {
-		data := *rawblobs[i]
-		refinedData, err := SerializeBlob(data)
-		if err != nil {
-			return nil, fmt.Errorf("Index %v: %v", i, err)
-		}
-		serialisedData = append(serialisedData, refinedData...)
-	}
+func isSkipEvm(indicator byte) bool {
```

Comment:
Can you add a comment on how this should work? 
Bitwise operations are not always the most intuitive to read and we will likely have readers wondering how this works.